src/scipp/interpolate/__init__.py

```diff
@@ -8,17 +8,29 @@
 """
 
 from ..core import array, Variable, DataArray, DimensionError, UnitError
-from ..core import irreducible_mask
+from ..core import dtype, irreducible_mask
 from ..compat.wrapping import wrap1d
 
-from typing import Callable
+from typing import Any, Callable, Union
 import uuid
 
+import numpy as np
+
+
+def _as_interpolation_type(x):
+    if isinstance(x, np.ndarray):
+        if x.dtype.kind == 'M':
+            return x.astype('int64', copy=False)
+    else:
+        if x.dtype == dtype.datetime64:
+            return x.astype('int64', copy=False)
+    return x
+
 
 def _midpoints(var, dim):
     a = var[dim, :-1]
     b = var[dim, 1:]
-    return a + 0.5 * (b - a)
+    return _as_interpolation_type(a) + 0.5 * (b - a)
```

Comment:
So this means datetime64 is now converted to `double`? Are we sure that this is correct? I see you added a unit test with increments of `1`, in which case `double` looks like the correct choice.... but what about rounding errors?

Say we operate in `ns`, at what point does the processing in `float64` cause rounding errors, which we would not have had in `int64`? It looks like not much before something like `1e19 ns` for `b - a`, but can you check? Is there any way we can get in the danger zone with a real example?

Reply:
I think we only get a problem when `a` and `b` are different by many orders of magnitude. If both are large, `b-a` can be large but it does not cause rounding errors.
In nanoseconds, we can go up to approx the year 2116. So for example, assuming `b=2116-01-01`, `a1=1970-01-01` and `a2=1970-02-01`, we get `b-a[1/2] ~ 2e18`. And the midpoints (using floats) are `a1/b ~ 2043-01-25` and `a2/b ~ 2043-02-10` meaning there is a difference of about 15 days which is half a month and correct.
You can of course require greater precision. But given a span of centuries as in the example, you should not expect a precision of better than days.

In particular, we are interpolating. If a 1e-16 or so relative difference in the independent coord matters, your interpolation is unstable and you shouldn't use it anyway.

The reason I went for floating point numbers is that with ints, the outputs are concatenated to whole numbers. So the interpolation would be evaluated not at the midpoint but at the int closest to it. In the extreme case of length 1 bins, this would 'interpolate' by taking the value at the lower edge. By using floats, we still get the middle.

Reply:
👍